File: frontend/src/services/annotations.ts
import api from './api'

export type AnnotationTargetType = 'content' | 'event' | 'atom' | 'event_pair' | 'atom_relation'

export interface AnnotationLabelItem {
  id: string
  task_id: string
  task_type: string
  target_type: string
  target_id: string
  label_payload: { value?: string; [key: string]: unknown }
  note?: string | null
  confidence?: number | null
  annotator: string
  supersedes_id?: string | null
  task_status: string
  created_at?: string | null
}

export interface AnnotationTaskItem {
  id: string
  task_type: string
  target_type: string
  target_id: string
  secondary_target_id?: string | null
  schema_version: string
  status: string
  priority: number
  reason?: string | null
  context_snapshot: Record<string, unknown>
  prediction_snapshot: Record<string, unknown>
  source_dataset?: string | null
  latest_label?: AnnotationLabelItem | null
  label_count: number
  created_at?: string | null
  updated_at?: string | null
}

export interface AnnotationStats {
  pending: number
  needs_adjudication: number
  labeled: number
  adjudicated: number
  total: number
  by_task_type: Record<string, number>
}

export interface SubmitAnnotationLabel {
  task_type: string
  target_type: AnnotationTargetType
  target_id: string
  secondary_target_id?: string
  schema_version?: string
  label_payload: { value: string; [key: string]: unknown }
  note?: string
  confidence?: number
  annotator?: string
  context_snapshot?: Record<string, unknown>
  prediction_snapshot?: Record<string, unknown>
  independent?: boolean
}

export const annotationsApi = {
  submitLabel: async (body: SubmitAnnotationLabel): Promise<AnnotationLabelItem> => {
    const response = await api.post<AnnotationLabelItem>('/annotations/labels', body)
    return response.data
  },

  getTarget: async (targetType: AnnotationTargetType, targetId: string): Promise<AnnotationTaskItem[]> => {
    const response = await api.get<{ items: AnnotationTaskItem[] }>(
      `/annotations/targets/${targetType}/${encodeURIComponent(targetId)}`,
    )
    return response.data.items
  },

  getStats: async (): Promise<AnnotationStats> => {
    const response = await api.get<AnnotationStats>('/annotations/stats')
    return response.data
  },

  getReviewQueue: async (limit = 100): Promise<{ items: AnnotationTaskItem[]; total: number }> => {
    const response = await api.get<{ items: AnnotationTaskItem[]; total: number }>(
      '/annotations/review-queue',
      { params: { status: 'actionable', limit } },
    )
    return response.data
  },

  adjudicate: async (
    taskId: string,
    value: string,
    rationale: string,
  ): Promise<void> => {
    await api.post(`/annotations/tasks/${taskId}/adjudicate`, {
      final_payload: { value },
      rationale,
      adjudicator: 'local-user',
      gold_candidate: true,
    })
  },
}

File: backend/app/schemas/annotations.py
"""API contracts for inline annotation and exceptional adjudication."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


AnnotationTargetType = Literal["content", "event", "atom", "event_pair", "atom_relation"]


class AnnotationLabelCreate(BaseModel):
    task_type: str = Field(min_length=1, max_length=48)
    target_type: AnnotationTargetType
    target_id: str = Field(min_length=1, max_length=128)
    secondary_target_id: str | None = Field(default=None, max_length=128)
    schema_version: str = Field(default="v1", min_length=1, max_length=24)
    label_payload: dict[str, Any]
    note: str | None = Field(default=None, max_length=4000)
    confidence: float | None = Field(default=None, ge=0, le=1)
    annotator: str = Field(default="local-user", min_length=1, max_length=128)
    context_snapshot: dict[str, Any] = Field(default_factory=dict)
    prediction_snapshot: dict[str, Any] = Field(default_factory=dict)
    independent: bool = False


class AnnotationLabelItem(BaseModel):
    id: str
    task_id: str
    task_type: str
    target_type: str
    target_id: str
    label_payload: dict[str, Any]
    note: str | None = None
    confidence: float | None = None
    annotator: str
    supersedes_id: str | None = None
    task_status: str
    created_at: str | None = None


class AnnotationTaskItem(BaseModel):
    id: str
    task_type: str
    target_type: str
    target_id: str
    secondary_target_id: str | None = None
    schema_version: str
    status: str
    priority: float
    reason: str | None = None
    context_snapshot: dict[str, Any] = Field(default_factory=dict)
    prediction_snapshot: dict[str, Any] = Field(default_factory=dict)
    source_dataset: str | None = None
    review_bucket: str | None = None
    labels: list[AnnotationLabelItem] = Field(default_factory=list)
    latest_label: AnnotationLabelItem | None = None
    label_count: int = 0
    created_at: str | None = None
    updated_at: str | None = None


class TargetAnnotationsResponse(BaseModel):
    target_type: str
    target_id: str
    items: list[AnnotationTaskItem] = Field(default_factory=list)


class AnnotationReviewQueueResponse(BaseModel):
    items: list[AnnotationTaskItem] = Field(default_factory=list)
    total: int
    bucket_counts: dict[str, int] = Field(default_factory=dict)


class AnnotationAdjudicationCreate(BaseModel):
    final_payload: dict[str, Any]
    rationale: str = Field(min_length=1, max_length=4000)
    adjudicator: str = Field(default="local-user", min_length=1, max_length=128)
    gold_candidate: bool = True


class AnnotationAdjudicationItem(BaseModel):
    id: str
    task_id: str
    final_payload: dict[str, Any]
    rationale: str
    adjudicator: str
    gold_candidate: bool
    created_at: str | None = None


class AnnotationStatsResponse(BaseModel):
    pending: int = 0
    needs_adjudication: int = 0
    labeled: int = 0
    adjudicated: int = 0
    retracted: int = 0
    total: int = 0
    by_task_type: dict[str, int] = Field(default_factory=dict)
    central_review: int = 0
    taxonomy_migration: int = 0
    deferred: int = 0
